pkg/volume/csi/csi_client.go

```diff
@@ -0,0 +1,182 @@
+		volumeid string,
+		readOnly bool,
+		targetPath string,
+		accessMode api.PersistentVolumeAccessMode,
+		fsType string,
+	) error
+	NodeUnpublishVolume(ctx grpctx.Context, volId string, targetPath string) error
+}
+
+// csiClient encapsulates all csi-plugin methods
+type csiDriverClient struct {
+	network    string
+	addr       string
+	conn       *grpc.ClientConn
+	idClient   csipb.IdentityClient
+	nodeClient csipb.NodeClient
+}
+
+func newCsiDriverClient(network, addr string) *csiDriverClient {
+	return &csiDriverClient{network: network, addr: addr}
+}
+
+// assertConnection ensures a valid connection has been established
+// if not, it creates a new connection and associated clients
+func (c *csiDriverClient) assertConnection() error {
+	if c.conn == nil {
+		conn, err := grpc.Dial(
+			c.addr,
+			grpc.WithInsecure(),
+			grpc.WithDialer(func(target string, timeout time.Duration) (net.Conn, error) {
+				return net.Dial(c.network, target)
+			}),
+		)
+		if err != nil {
+			return err
+		}
+		c.conn = conn
+		c.idClient = csipb.NewIdentityClient(conn)
+		c.nodeClient = csipb.NewNodeClient(conn)
+	}
+
+	return nil
+}
+
+// AssertCSIVersion determines if driver supports specified version
+func (c *csiDriverClient) AssertSupportedVersion(ctx grpctx.Context, ver *csipb.Version) error {
+	if err := c.assertConnection(); err != nil {
+		return err
+	}
+
+	rsp, err := c.idClient.GetSupportedVersions(ctx, &csipb.GetSupportedVersionsRequest{})
+	if err != nil {
+		return err
+	}
+	supported := false
+	vers := rsp.GetSupportedVersions()
+	for _, v := range vers {
+		if v.GetMajor() == ver.GetMajor() &&
+			v.GetMinor() == ver.GetMinor() &&
+			v.GetPatch() == ver.GetPatch() {
+			supported = true
+		}
+	}
+	if !supported {
+		return fmt.Errorf("version %d.%d.%d not supported", ver.GetMajor(), ver.GetMinor(), ver.GetPatch())
+	}
+	glog.V(4).Infof(log("version %d.%d.%d supported", ver.GetMajor(), ver.GetMinor(), ver.GetPatch()))
+	return nil
+}
+
+func (c *csiDriverClient) NodePublishVolume(
+	ctx grpctx.Context,
+	volId string,
+	readOnly bool,
+	targetPath string,
+	accessMode api.PersistentVolumeAccessMode,
+	fsType string,
+) error {
+
+	if volId == "" {
+		return errors.New("missing volume id")
+	}
+	if targetPath == "" {
+		return errors.New("missing target path")
+	}
+	if err := c.assertConnection(); err != nil {
+		glog.Errorf("%v: failed to assert a connection: %v", csiPluginName, err)
+		return err
+	}
+
+	req := &csipb.NodePublishVolumeRequest{
+		Version:           csiVersion,
+		VolumeId:          volId,
+		TargetPath:        targetPath,
+		Readonly:          readOnly,
+		PublishVolumeInfo: map[string]string{"device": "/dev/null"}, //TODO where is this from
+
+		VolumeCapability: &csipb.VolumeCapability{
+			AccessMode: &csipb.VolumeCapability_AccessMode{
+				Mode: asCSIAccessMode(accessMode),
+			},
+			AccessType: &csipb.VolumeCapability_Mount{
+				Mount: &csipb.VolumeCapability_MountVolume{
+					FsType: fsType,
```

Comment:
keep in mind that Erin Boyd's block storage project may need to enhance this part.